config/config.go

```diff
@@ -679,159 +212,73 @@ func GetDuration(key string, defaultValueInTimescaleUnits int64, timeScale time.
-}
-
-// GetEnvAsBool returns the boolean environment value stored in key variable
-func GetEnvAsBool(name string, defaultVal bool) bool {
-	valueStr := GetEnv(name, "")
-	if value, err := strconv.ParseBool(valueStr); err == nil {
-		return value
-	}
-
-	return defaultVal
-}
-
-// GetRequiredEnv returns the environment value stored in key variable, no default
-func GetRequiredEnv(key string) string {
-	if value, exists := os.LookupEnv(key); exists {
-		return value
-	}
-	panic(fmt.Errorf("fatal error, no required environment variable: %s", key))
-}
-
-// GetEnvErr returns the value of environment variable, error if not exists
-func GetEnvErr(key string) (string, error) {
-	if value, exists := os.LookupEnv(key); exists {
-		return value, nil
-	}
-	return "", fmt.Errorf("missing required environment variable: %s", key)
+// IsSet checks if config is set for a key
+func (c *Config) IsSet(key string) bool {
+	c.vLock.RLock()
+	defer c.vLock.RUnlock()
+	c.bindEnv(key)
+	return c.v.IsSet(key)
 }
 
 // Override Config by application or command line
 
-// SetBool override existing config
-func SetBool(key string, value bool) {
-	viper.Set(key, value)
-}
-
-func SetString(key, value string) {
-	viper.Set(key, value)
-}
-
-func SetHotReloadablesForcefully(key string, value interface{}) {
-	viper.Set(key, value)
-	configVarLock.RLock()
-	defer configVarLock.RUnlock()
-	_ = checkAndHotReloadConfig(hotReloadableConfig)
-}
-
-// GetWorkspaceToken returns the workspace token provided in the environment variables
-// Env variable CONFIG_BACKEND_TOKEN is deprecating soon
-// WORKSPACE_TOKEN is newly introduced. This will override CONFIG_BACKEND_TOKEN
-func GetWorkspaceToken() string {
-	token := GetEnv("WORKSPACE_TOKEN", "")
-	if token != "" && token != "<your_token_here>" {
-		return token
-	}
-	return GetEnv("CONFIG_BACKEND_TOKEN", "")
-}
-
-func GetNamespaceIdentifier() string {
-	k8sNamespace := GetKubeNamespace()
-	if k8sNamespace != "" {
-		return k8sNamespace
+// Set override existing config
+func Set(key string, value interface{}) {
+	c.Set(key, value)
+}
+
+// Set override existing config
+func (c *Config) Set(key string, value interface{}) {
+	c.vLock.Lock()
+	c.v.Set(key, value)
+	c.vLock.Unlock()
+	c.onConfigChange()
+}
+
+// bindEnv handles rudder server's unique snake case replacement by registering
+// the environment variables to viper, that would otherwise be ignored.
+// Viper uppercases keys before sending them to its EnvKeyReplacer, thus
+// the replacer cannot detect camelCase keys.
+func (c *Config) bindEnv(key string) {
+	envVar := key
+	if !upperCaseMatch.MatchString(key) {
+		envVar = ConfigKeyToEnvRudder(key)
+	}
+	// bind once
+	c.envLock.RLock()
+	if _, ok := c.envs[key]; !ok {
+		c.envLock.RUnlock()
+		c.envLock.Lock() // don't really care about race here, setting the same value
+		c.envs[strings.ToUpper(key)] = envVar
+		c.envLock.Unlock()
+	} else {
+		c.envLock.RUnlock()
```

Comment:
[optional] Even though you don't care about the race it can make sense to simplify this bit and just `Lock()` and `defer Unlock()` instead of mixing.

Reply:
For a strange reason I tend to prefer this kind of (biased) optimisation, considering that in most cases the key will be already in the map, so we don't really need to be pessimistic and block competing config readers.

Reply:
Yes I understand, I'm suggesting mostly to simplify the logic. I don't see that much of a performance impact in this case so perhaps simplyfing can be good. Up to you of course :+1: